ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,435 @@
+		pm.isNewProfile = false
+	} else if err == ipn.ErrStateNotExist {
+		pm.prefs = emptyPrefs
+		pm.isNewProfile = true
+	} else {
+		return err
+	}
+	pm.currentUserID = uid
+	return nil
+}
+
+func (pm *ProfileManager) findProfileByKey(key ipn.StateKey) *loginProfile {
+	for _, p := range pm.knownProfiles {
+		if p.Key == key {
+			return p
+		}
+	}
+	return nil
+}
+
+func (pm *ProfileManager) setUnattendedModeAsConfigured() error {
+	if pm.currentUserID == "" {
+		return nil
+	}
+
+	if pm.prefs.ForceDaemon() {
+		return pm.store.WriteState(ipn.ServerModeStartKey, []byte(pm.currentProfile.Key))
+	} else {
+		return pm.store.WriteState(ipn.ServerModeStartKey, nil)
+	}
+}
+
+// Reset unloads the current profile, if any.
+func (pm *ProfileManager) Reset() {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.prefs = emptyPrefs
+	pm.currentUserID = ""
+	pm.currentProfile = nil
+}
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+func randID() string {
+	var b [4]byte
+	rand.Read(b[:])
+	return fmt.Sprintf("%x", b)
+}
+
+// SetPrefs sets the current profile's prefs to the provided value.
+// It also saves the prefs to the StateStore. It stores a copy of the
+// provided prefs, which may be accessed via CurrentPrefs.
+func (pm *ProfileManager) SetPrefs(prefsIn ipn.PrefsView) error {
+	prefs := prefsIn.AsStruct().View()
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	ps := prefs.Persist()
+	if !pm.isNewProfile || ps == nil || ps.LoginName == "" {
+		return pm.setPrefsLocked(prefs)
+	}
+	id, k := newUnusedKey(pm.knownProfiles)
+	pm.currentProfile = &loginProfile{
+		Name:        ps.LoginName,
+		Key:         k,
+		ID:          id,
+		LocalUserID: pm.currentUserID,
+	}
+	pm.knownProfiles[ps.LoginName] = pm.currentProfile
+	if err := pm.writeKnownProfiles(); err != nil {
+		delete(pm.knownProfiles, ps.LoginName)
+		return err
+	}
+	if err := pm.setAsUserSelectedProfileLocked(); err != nil {
+		return err
+	}
+	if err := pm.setPrefsLocked(prefs); err != nil {
+		return err
+	}
+	pm.isNewProfile = false
+	return nil
+}
+
+func newUnusedKey(knownProfiles map[string]*loginProfile) (id string, key ipn.StateKey) {
+keyGenLoop:
+	for {
+		id := randID()
+		for _, kp := range knownProfiles {
+			if kp.ID == id {
+				continue keyGenLoop
+			}
+		}
+		return id, ipn.StateKey("profile-" + id)
+	}
+}
+
+func (pm *ProfileManager) setPrefsLocked(clonedPrefs ipn.PrefsView) error {
+	pm.prefs = clonedPrefs
+	if pm.currentProfile == nil {
+		return nil
+	}
+	if err := pm.writePrefsToStore(pm.currentProfile.Key, pm.prefs); err != nil {
+		return err
+	}
+	return pm.setUnattendedModeAsConfigured()
+}
+
+func (pm *ProfileManager) writePrefsToStore(key ipn.StateKey, prefs ipn.PrefsView) error {
+	if key == "" {
+		return nil
+	}
+	if err := pm.store.WriteState(key, prefs.ToBytes()); err != nil {
+		pm.logf("WriteState(%q): %v", key, err)
+		return err
+	}
+	return nil
+}
+
+// Profiles returns the list of known profiles.
+func (pm *ProfileManager) Profiles() []string {
```

Comment:
Returning a struct with the `DisplayName`, `ProfilePicURL`, etc. fields would also be good for the Mac UI.